explorer/chartscache.go

```diff
@@ -124,3 +126,278 @@ func ReadCacheFile(filePath string, height int64) error {
 
 	return nil
 }
+
+// ChainMonitor defines data needed to handle cache data reorganization.
+type ChainMonitor struct {
+	ctx       context.Context
+	wg        *sync.WaitGroup
+	explorer  *explorerUI
+	cacheChan chan *txhelpers.ReorgData
+}
+
+// NewCacheChainMonitor returns an initialized charts cache chain monitor instance.
+func (exp *explorerUI) NewCacheChainMonitor(ctx context.Context, wg *sync.WaitGroup,
+	reorgChan chan *txhelpers.ReorgData) *ChainMonitor {
+	return &ChainMonitor{
+		ctx:       ctx,
+		wg:        wg,
+		explorer:  exp,
+		cacheChan: reorgChan,
+	}
+}
+
+// ReorgHandler handles the charts cache data reorganization.
+func (m *ChainMonitor) ReorgHandler() {
+	defer m.wg.Done()
+	for {
+		select {
+		case data, ok := <-m.cacheChan:
+			if !ok {
+				return
+			}
+
+			// Drop all entries since since the ancestor block after reorganization
+			// of all PG and Sqlite DB is complete.
+			if err := m.explorer.DropToAncestor(uint64(data.NewChainHeight)); err != nil {
+				log.Errorf("cache reorg failed: %v", err)
+				return
+			}
+
+			// Initiate cache update after drop the entries till the ancestor bloxk.
+			m.explorer.prePopulateChartsData()
+
+			data.WG.Done()
+
+		case <-m.ctx.Done():
+			return
+		}
+	}
+}
+
+// DropToAncestor drops all the chart data entries since the common ancestor
+// block. If the ancestor block is in the last entry added, then the last chart
+// data entry is the only one that is dropped.
+func (exp *explorerUI) DropToAncestor(heightAfterAncestor uint64) error {
+	timeAfterAncestor, err := exp.explorerSource.BlockTimeByHeight(int64(heightAfterAncestor))
+	if err != nil {
+		return err
+	}
+
+	cacheData := cacheChartsData.get()
+
+	// "avg-block-size" chart data.
+	data, ok := cacheData[dbtypes.AvgBlockSize]
+	if ok {
+		var index = len(data.Time) - 1
+		for ; index > 0; index-- {
+			if data.Time[index].UNIX() < timeAfterAncestor {
+				break
+			}
+		}
+		cacheData[dbtypes.AvgBlockSize].Time = cacheData[dbtypes.AvgBlockSize].Time[:index]
+		cacheData[dbtypes.AvgBlockSize].Size = cacheData[dbtypes.AvgBlockSize].Size[:index]
+	}
+
+	// "blockchain-size" chart data.
+	data, ok = cacheData[dbtypes.BlockChainSize]
+	if ok {
+		var index = len(data.Time) - 1
+		for ; index > 0; index-- {
+			if data.Time[index].UNIX() < timeAfterAncestor {
+				break
+			}
+		}
+		cacheData[dbtypes.BlockChainSize].Time = cacheData[dbtypes.BlockChainSize].Time[:index]
+		cacheData[dbtypes.BlockChainSize].ChainSize = cacheData[dbtypes.BlockChainSize].ChainSize[:index]
```

Comment:
Given the indexing and number of edge case, the above loop and slice resize should be testable.  Having it be reusable would also benefit a number of the other data sets.

Reply:
> above loop and slice resize

Since `cacheData` is now a map are you referring to the individual chart data sets which are the only slices here?

Reply:
More abstract than charts data an the map.
Two things: (1) computing `index` given a slice of times and another reference time, and (2) reslicing an arbitrary slice with the index check the slice length to avoid panic and then reslice.